store/bench_test.go

```diff
@@ -0,0 +1,378 @@
+	has    int
+	get    int
+	set    int
+	delete int
+}
+
+type counts struct {
+	has    int
+	get    int
+	set    int
+	delete int
+}
+
+func generateGradedPercentages() []percentages {
+	var sampledPercentages []percentages
+	sampleX := percentages{has: 2, get: 55, set: 40, delete: 3}
+	sampledPercentages = append(sampledPercentages, sampleX)
+	for a := 0; a < 100; a += 20 {
+		for b := 0; b <= 100-a; b += 20 {
+			for c := 0; c < 100-a-b; c += 20 {
+				sample := percentages{
+					has:    a,
+					get:    b,
+					set:    c,
+					delete: 100 - a - b - c,
+				}
+				sampledPercentages = append(sampledPercentages, sample)
+			}
+		}
+	}
+	return sampledPercentages
+}
+
+func generateExtremePercentages() []percentages {
+	return []percentages{
+		{100, 0, 0, 0},
+		{0, 100, 0, 0},
+		{0, 0, 100, 0},
+		{0, 0, 0, 100},
+	}
+}
+
+type benchmark struct {
+	name        string
+	percentages percentages
+	dbType      tmdb.BackendType
+	counts      counts
+}
+
+func generateBenchmarks(dbBackendTypes []tmdb.BackendType, sampledPercentages []percentages, sampledCounts []counts) []benchmark {
+	var benchmarks []benchmark
+	for _, dbType := range dbBackendTypes {
+		if len(sampledPercentages) > 0 {
+			for _, p := range sampledPercentages {
+				name := fmt.Sprintf("r-%s-%d-%d-%d-%d", dbType, p.has, p.get, p.set, p.delete)
+				benchmarks = append(benchmarks, benchmark{name: name, percentages: p, dbType: dbType, counts: counts{}})
+			}
+		} else if len(sampledCounts) > 0 {
+			for _, c := range sampledCounts {
+				name := fmt.Sprintf("d-%s-%d-%d-%d-%d", dbType, c.has, c.get, c.set, c.delete)
+				benchmarks = append(benchmarks, benchmark{name: name, percentages: percentages{}, dbType: dbType, counts: c})
+			}
+		}
+	}
+	return benchmarks
+}
+
+type store interface {
+	Has(key []byte) bool
+	Get(key []byte) []byte
+	Set(key []byte, value []byte)
+	Delete(key []byte)
+	Commit() types.CommitID
+}
+
+type storeV2 struct {
+	*storev2.Store
+	storev2types.KVStore
+}
+
+func simpleStoreConfig() (storev2.StoreConfig, error) {
+	opts := storev2.DefaultStoreConfig()
+	err := opts.RegisterSubstore(skey_1.Name(), types.StoreTypePersistent)
+	if err != nil {
+		return storev2.StoreConfig{}, err
+	}
+	return opts, nil
+}
+
+func sampleOperation(p percentages) string {
+	ops := []string{"Has", "Get", "Set", "Delete"}
+	thresholds := []int{p.has, p.has + p.get, p.has + p.get + p.set}
+	r := rand.Intn(100)
+	for i := 0; i < len(thresholds); i++ {
+		if r < thresholds[i] {
+			return ops[i]
+		}
+	}
+	return ops[3]
+}
+
+func runRandomizedOperations(b *testing.B, s store, totalOpsCount int, p percentages) {
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		for j := 0; j < totalOpsCount; j++ {
+			b.StopTimer()
+			op := sampleOperation(p)
+			b.StartTimer()
+
+			switch op {
+			case "Has":
+				s.Has(randBytes(12))
+			case "Get":
+				s.Get(randBytes(12))
+			case "Set":
+				s.Set(randBytes(12), randBytes(50))
+			case "Delete":
+				s.Delete(randBytes(12))
+			}
```

Comment:
If a benchmark uses random (nondeterministic) data like this, then each execution performs different work. But the expectation is that each loop is doing exactly the same thing. If that's not the case, the benchmark doesn't produce accurate results.

If you want to use randomized data, create a rand.Source from a seed int64 that's random by default but can be provided by the caller if they so choose. And be sure to print the seed, so that callers can re-run exactly the same scenario. Then, generate all of the random data _before_ entering any b.N loops.

Reply:
Yup. That's pretty much how sims work. I support this idea.